File: python/question1_tests.py
import random
from test_harness import TestHarness, Case


def build_harness(hash_map_class):
    """Build a test harness for the custom HashMap implementation."""

    def _new():
        return hash_map_class()

    # -- Correctness tests ------------------------------------

    def test_basic_put_get():
        m = _new()
        m.put(1, 100)
        m.put(2, 200)
        m.put(3, 300)
        return m.get(1) == 100 and m.get(2) == 200 and m.get(3) == 300

    def test_get_missing():
        m = _new()
        m.put(1, 100)
        return m.get(999) == -1 and m.get(0) == -1

    def test_overwrite():
        m = _new()
        m.put(1, 100)
        m.put(1, 999)
        return m.get(1) == 999

    def test_contains_key():
        m = _new()
        m.put(42, 1)
        return m.contains_key(42) and not m.contains_key(99)

    def test_remove_existing():
        m = _new()
        m.put(5, 50)
        removed = m.remove(5)
        return removed and m.get(5) == -1 and not m.contains_key(5)

    def test_remove_non_existent():
        m = _new()
        m.put(1, 10)
        return not m.remove(42)

    def test_tombstone_recovery():
        m = _new()
        m.put(5, 50)
        m.remove(5)
        m.put(5, 999)
        return m.get(5) == 999 and m.contains_key(5) and m.size() == 1

    def test_triggers_resize():
        m = _new()
        for i in range(50):
            m.put(i * 37, i)
        for i in range(50):
            if m.get(i * 37) != i:
                return False
        return m.size() == 50

    def test_negative_keys():
        m = _new()
        m.put(-1, 10)
        m.put(-100, 20)
        m.put(-2147483648, 30)
        return (m.get(-1) == 10
                and m.get(-100) == 20
                and m.get(-2147483648) == 30
                and m.contains_key(-1))

    def test_size_accuracy():
        m = _new()
        for i in range(5):
            m.put(i, i * 10)
        m.remove(1)
        m.remove(3)
        return m.size() == 3

    def test_resize_after_removals():
        m = _new()
        for i in range(20):
            m.put(i, i * 10)
        for i in range(5):
            m.remove(i)
        for i in range(20, 50):
            m.put(i, i * 10)
        if m.size() != 45:
            return False
        for i in range(5, 50):
            if m.get(i) != i * 10:
                return False
        return True

    def test_many_entries():
        m = _new()
        for i in range(1000):
            m.put(i, i * 7)
        for i in range(1000):
            if m.get(i) != i * 7:
                return False
        return m.size() == 1000

    # -- Benchmark helper -------------------------------------

    def bench_100k():
        m = _new()
        for i in range(100_000):
            m.put(i, i)
        total = 0
        for i in range(100_000):
            total += m.get(i)
        return total > 0

    # -- Scaling helper ---------------------------------------

    def gen_keys(n):
        gen = random.Random(42)
        return [gen.randint(0, n * 10) for _ in range(n)]

    def solve_scaling(keys):
        m = _new()
        for k in keys:
            m.put(k, k)
        for k in keys:
            m.get(k)
        return True

    # -- Random stress test generator -------------------------

    rng = random.Random(42)

    def case_generator():
        num_ops = rng.randint(20, 50)
        ops = []
        for _ in range(num_ops):
            op_type = rng.randint(0, 4)
            key = rng.randint(-50, 149)
            value = rng.randint(0, 999)
            ops.append((op_type, key, value))

        # Compute reference results using built-in dict
        ref = {}
        expected_results = []
        for op_type, key, value in ops:
            if op_type == 0:
                ref[key] = value
                expected_results.append(None)
            elif op_type == 1:
                expected_results.append(ref.get(key, -1))
            elif op_type == 2:
                was_present = key in ref
                if was_present:
                    del ref[key]
                expected_results.append(was_present)
            elif op_type == 3:
                expected_results.append(key in ref)
            elif op_type == 4:
                expected_results.append(len(ref))

        ops_snapshot = list(ops)
        exp_snapshot = list(expected_results)

        def fn():
            m = _new()
            for i, (op_type, key, value) in enumerate(ops_snapshot):
                if op_type == 0:
                    m.put(key, value)
                elif op_type == 1:
                    if m.get(key) != exp_snapshot[i]:
                        return False
                elif op_type == 2:
                    if m.remove(key) != exp_snapshot[i]:
                        return False
                elif op_type == 3:
                    if m.contains_key(key) != exp_snapshot[i]:
                        return False
                elif op_type == 4:
                    if m.size() != exp_snapshot[i]:
                        return False
            return True

        return Case(f"{num_ops} random ops", fn, True)

    # -- Build harness ----------------------------------------

    return TestHarness("Custom HashMap (Open Addressing)") \
        .add_test("Basic put/get", test_basic_put_get, True) \
        .add_test("Get missing key returns -1", test_get_missing, True) \
        .add_test("Overwrite value", test_overwrite, True) \
        .add_test("containsKey true/false", test_contains_key, True) \
        .add_test("Remove existing key", test_remove_existing, True) \
        .add_test("Remove non-existent key returns false", test_remove_non_existent, True) \
        .add_test("Tombstone recovery: remove then re-put", test_tombstone_recovery, True) \
        .add_test("Triggers resize (more than initial capacity)", test_triggers_resize, True) \
        .add_test("Negative keys", test_negative_keys, True) \
        .add_test("Size accuracy (put 5, remove 2)", test_size_accuracy, True) \
        .add_test("Resize after removals preserves data", test_resize_after_removals, True) \
        .add_test("1000 entries: put then get all", test_many_entries, True) \
        .add_benchmark("100,000 put + get operations", bench_100k) \
        .add_scaling("N (put + get operations)",
                     [1_000, 5_000, 10_000, 50_000, 100_000],
                     gen_keys, solve_scaling) \
        .add_random_tests("vs built-in dict", 1000, case_generator)

File: java/Question1_Tests.java
import java.util.*;
import java.util.function.*;

public class Question1_Tests {

    private static int[] genKeys(int n) {
        Random gen = new Random(42);
        int[] keys = new int[n];
        for (int i = 0; i < n; i++) {
            keys[i] = gen.nextInt(n * 10);
        }
        return keys;
    }

    public static TestHarness<Boolean> buildHarness() {
        Random rng = new Random(42);

        return TestHarness.<Boolean>forProblem("Custom HashMap (Open Addressing)")

            // ── Correctness tests ───────────────────────────

            .addTest("Basic put/get", () -> {
                MyHashMap m = new MyHashMap();
                m.put(1, 100);
                m.put(2, 200);
                m.put(3, 300);
                return m.get(1) == 100 && m.get(2) == 200 && m.get(3) == 300;
            }, true)

            .addTest("Get missing key returns -1", () -> {
                MyHashMap m = new MyHashMap();
                m.put(1, 100);
                return m.get(999) == -1 && m.get(0) == -1;
            }, true)

            .addTest("Overwrite value", () -> {
                MyHashMap m = new MyHashMap();
                m.put(1, 100);
                m.put(1, 999);
                return m.get(1) == 999;
            }, true)

            .addTest("containsKey true/false", () -> {
                MyHashMap m = new MyHashMap();
                m.put(42, 1);
                return m.containsKey(42) && !m.containsKey(99);
            }, true)

            .addTest("Remove existing key", () -> {
                MyHashMap m = new MyHashMap();
                m.put(5, 50);
                boolean removed = m.remove(5);
                return removed && m.get(5) == -1 && !m.containsKey(5);
            }, true)

            .addTest("Remove non-existent key returns false", () -> {
                MyHashMap m = new MyHashMap();
                m.put(1, 10);
                return !m.remove(42);
            }, true)

            .addTest("Tombstone recovery: remove then re-put", () -> {
                MyHashMap m = new MyHashMap();
                m.put(5, 50);
                m.remove(5);
                m.put(5, 999);
                return m.get(5) == 999 && m.containsKey(5) && m.size() == 1;
            }, true)

            .addTest("Triggers resize (more than initial capacity)", () -> {
                MyHashMap m = new MyHashMap();
                // Insert well past the initial capacity of 16
                for (int i = 0; i < 50; i++) {
                    m.put(i * 37, i);
                }
                for (int i = 0; i < 50; i++) {
                    if (m.get(i * 37) != i) {
                        return false;
                    }
                }
                return m.size() == 50;
            }, true)

            .addTest("Negative keys", () -> {
                MyHashMap m = new MyHashMap();
                m.put(-1, 10);
                m.put(-100, 20);
                m.put(Integer.MIN_VALUE, 30);
                return m.get(-1) == 10
                    && m.get(-100) == 20
                    && m.get(Integer.MIN_VALUE) == 30
                    && m.containsKey(-1);
            }, true)

            .addTest("Size accuracy (put 5, remove 2)", () -> {
                MyHashMap m = new MyHashMap();
                for (int i = 0; i < 5; i++) {
                    m.put(i, i * 10);
                }
                m.remove(1);
                m.remove(3);
                return m.size() == 3;
            }, true)

            .addTest("Resize after removals preserves data", () -> {
                MyHashMap m = new MyHashMap();
                for (int i = 0; i < 20; i++) {
                    m.put(i, i * 10);
                }
                for (int i = 0; i < 5; i++) {
                    m.remove(i);
                }
                // Add more to trigger further resizing
                for (int i = 20; i < 50; i++) {
                    m.put(i, i * 10);
                }
                if (m.size() != 45) {
                    return false;
                }
                for (int i = 5; i < 50; i++) {
                    if (m.get(i) != i * 10) {
                        return false;
                    }
                }
                return true;
            }, true)

            .addTest("1000 entries: put then get all", () -> {
                MyHashMap m = new MyHashMap();
                for (int i = 0; i < 1000; i++) {
                    m.put(i, i * 7);
                }
                for (int i = 0; i < 1000; i++) {
                    if (m.get(i) != i * 7) {
                        return false;
                    }
                }
                return m.size() == 1000;
            }, true)

            // ── Benchmark ───────────────────────────────────

            .addBenchmark("100,000 put + get operations", () -> {
                MyHashMap m = new MyHashMap();
                for (int i = 0; i < 100_000; i++) {
                    m.put(i, i);
                }
                int sum = 0;
                for (int i = 0; i < 100_000; i++) {
                    sum += m.get(i);
                }
                return sum > 0;
            })

            // ── Scaling analysis ────────────────────────────

            .addScaling("N (put + get operations)",
                new int[]{1_000, 5_000, 10_000, 50_000, 100_000},
                n -> genKeys(n),
                (int[] input) -> {
                    MyHashMap m = new MyHashMap();
                    for (int k : input) {
                        m.put(k, k);
                    }
                    for (int k : input) {
                        m.get(k);
                    }
                    return true;
                })

            // ── Random stress tests ─────────────────────────

            .addRandomTests("vs java.util.HashMap", 1000, () -> {
                int numOps = 20 + rng.nextInt(31);
                int[] opTypes = new int[numOps];
                int[] opKeys = new int[numOps];
                int[] opVals = new int[numOps];
                for (int i = 0; i < numOps; i++) {
                    opTypes[i] = rng.nextInt(5);
                    opKeys[i] = rng.nextInt(200) - 50;
                    opVals[i] = rng.nextInt(1000);
                }

                // Compute reference results using java.util.HashMap
                HashMap<Integer, Integer> ref = new HashMap<>();
                int[] refGet = new int[numOps];
                boolean[] refBool = new boolean[numOps];
                int[] refSize = new int[numOps];
                for (int i = 0; i < numOps; i++) {
                    switch (opTypes[i]) {
                        case 0:
                            ref.put(opKeys[i], opVals[i]);
                            break;
                        case 1:
                            refGet[i] = ref.getOrDefault(opKeys[i], -1);
                            break;
                        case 2:
                            refBool[i] = ref.remove(opKeys[i]) != null;
                            break;
                        case 3:
                            refBool[i] = ref.containsKey(opKeys[i]);
                            break;
                        case 4:
                            refSize[i] = ref.size();
                            break;
                    }
                }

                return new TestHarness.Case<>(
                    numOps + " random ops",
                    () -> {
                        MyHashMap m = new MyHashMap();
                        for (int i = 0; i < numOps; i++) {
                            switch (opTypes[i]) {
                                case 0:
                                    m.put(opKeys[i], opVals[i]);
                                    break;
                                case 1:
                                    if (m.get(opKeys[i]) != refGet[i]) {
                                        return false;
                                    }
                                    break;
                                case 2:
                                    if (m.remove(opKeys[i]) != refBool[i]) {
                                        return false;
                                    }
                                    break;
                                case 3:
                                    if (m.containsKey(opKeys[i]) != refBool[i]) {
                                        return false;
                                    }
                                    break;
                                case 4:
                                    if (m.size() != refSize[i]) {
                                        return false;
                                    }
                                    break;
                            }
                        }
                        return true;
                    },
                    true
                );
            });
    }
}
